### ISSUE TITLE
when launching a background process from fish_prompt, prompt feels slower than `time` reports

### ISSUE THREAD

**Issue by IlanCosman** [fish-shell/fish-shell#7559]
```fish
function fish_prompt
    sleep 5 &
    echo $PWD
end
```

![Peek 2020-12-18 19-56](https://user-images.githubusercontent.com/6318351/102680231-45020b00-416b-11eb-9c7d-ab7b52b9439d.gif)

I looked over the open issues and couldn't find anything like this, but maybe I missed it. `fish 3.1.2`

**Comment by IlanCosman** [comment]
The plot thickens. It seems to be somewhat based on the amount of time the backgrounded task takes.

![faster](https://user-images.githubusercontent.com/6318351/102680577-61ec0d80-416e-11eb-9f06-a37097b934a4.gif)

**Comment by ridiculousfish** [comment]
Ok, this is saying that when you launch a background process from fish_prompt, then the prompt "feels" slower than `time` reports. But drawing a prompt is more than just running `fish_prompt` so this is in some sense expected.

It may also be the case that we're spending a lot of time reaping those background processes since we have to `waitpid` for each one. You may also be hitting what e334becefbf221db19e8e4e4c550da036db3b61b was supposed to address (not sure if there's an attached issue) where we end up with too many threads.  Can you please use `sample` (on Mac) or `perf` on Linux to collect a trace, and attach it?

**Comment by IlanCosman** [comment]
> Ok, this is saying that when you launch a background process from fish_prompt, then the prompt "feels" slower than time reports. 

Thanks for the much better summary 😂

> But drawing a prompt is more than just running fish_prompt so this is in some sense expected.

I didn't know that drawing the prompt involved much more than running `fish_prompt`, so it wasn't expected for me. Thanks for the info 😄

> Can you please use sample (on Mac) or perf on Linux to collect a trace, and attach it?

Absolutely! I installed `perf`, but I have no idea how to use it. I assume something like `sudo perf trace`, but that just spams my terminal. Could you give me some guidance?

**Comment by ridiculousfish** [comment]
Hmm, try this, in another terminal window:

    perf record -F 99 --call-graph dwarf -p 12345 -g -- sleep 10

where 12345 is the pid of the running fish instance. That gives you 10 seconds to repro the problem.

After you can run `perf report -n --stdio` to get a report of what fish was doing.

**Comment by IlanCosman** [comment]
Hope I did this right, and that this is what you were looking for 😄

[perf_trace.txt](https://github.com/fish-shell/fish-shell/files/5720501/perf_trace.txt)

**Comment by ridiculousfish** [comment]
Looks like that only got one sample - sorry, by "reproduce the problem" I mean reproduce it as much as you can during those ten seconds (e.g. hold down enter key or whatever). Unfortunately I'm not a `perf` expert, you can [see some of the examples here](http://www.brendangregg.com/perf.html#OneLiners).

When I try to repro on my Mac and produce a lot of background processes, I see a fair amount of time spent in `waitpid` during `fish_title`, i.e. we're checking to see if the background processes are finished. That's consistent with your observation that shorter times make the problem less noticeable, because there would be fewer  background procs. Probably the stack traces would show this.

**Comment by IlanCosman** [comment]
I also have access to a Mac if you know `sample` better. Heres my second attempt at a `perf` trace.

[perf_trace.txt](https://github.com/fish-shell/fish-shell/files/5721402/perf_trace.txt)

It sounds like you're able to reproduce the problem on your side though.

**Comment by IlanCosman** [comment]
@ridiculousfish I see you marked this as fish-future. That makes sense given that this isn't a high-priority issue for most people--what kinda weirdo starts background processes in their `fish_prompt` anyway? 😂 

Unfortunately this is a big deal for my project, so can I ask you to provide a bit of guidance on this? Is this something I could reasonably attempt as a novice, or would I be totally out of my depth here? Do you have any idea as to where I would start?

**Comment by ridiculousfish** [comment]
I'll take another look as soon as I can. Perhaps we can just defer reaping background processes from `fish_title` and other interactive functions.

**Comment by ridiculousfish** [comment]
@IlanCosman can you check how many jobs are running when it reproduces in the slow case? `count (jobs)` will report that. Might be good to run that as part of `fish_prompt` instead of after, to avoid the race.

**Comment by IlanCosman** [comment]
> check how many jobs are running when it reproduces in the slow case

Nothing unexpected here. Just 1 new job per `Enter` press. It's slow right off the bat though, so this particular issue is clearly not about the number of background jobs. 

Prompt speed is definitely effected by the number though. Number-of-jobs slowdown is much less dramatic than starting-background-process-in-`fish_prompt` slowdown though. `time` is also able to capture the slowdown here.
```fish
ilan@arch ~> time fish_prompt
ilan@arch ~> 
________________________________________________________
Executed in    2.17 millis    fish           external 
   usr time    0.73 millis    0.00 micros  729.00 micros 
   sys time    1.41 millis  1412.00 micros    0.00 micros 

ilan@arch ~> for i in (seq 1000)
                 sleep 20 &
             end
ilan@arch ~> time fish_prompt
ilan@arch ~> 
________________________________________________________
Executed in   25.87 millis    fish           external 
   usr time    0.74 millis    0.00 millis  737.00 micros 
   sys time   24.99 millis   24.99 millis    0.00 micros 
```

**Comment by ridiculousfish** [comment]
Oh wow, it just clicked for me, it is indeed a fish bug. When we run `fish_prompt`, we redirect its output to a pipe, which we read from. Usually we stop reading on EOF, when the write end of the pipe is closed. But the write end of the pipe may escape, if it is inherited by a background process or daemon, so fish wakes up every 100 msec to check if the job is exited (even though the pipe is live).

So this adds up to 100 msec of latency ending the command substitution. If the bg job exits more quickly, then the write end of the pipe is closed and we continue; that explains why sleeping for < 100msec reduces the latency, but above 100msec it won't matter.

[Here's the big comment explaining it](https://github.com/fish-shell/fish-shell/blob/4116aaeb5fc90aa7088f3be608d08dc2a1f66583/src/io.cpp#L94). When I wrote that, I thought that spawning bg jobs from command substitutions wasn't something anyone cared about, and fish just had to not hang. Looks like we care now.

Probably we can remove this 100msec polling in 3.2. @IlanCosman a possible workaround for you on 3.1.2 is to ensure that your bg job doesn't inherit the cmdsub pipe, by redirecting or closing stdout. For example:

    function fish_prompt
         sleep 5 >&- &
         echo $PWD
    end

this will eliminate the extra latency.

**Comment by IlanCosman** [comment]
That's great to hear! Your workaround works perfectly too, no more annoying latency. Thanks a ton 😄

**Comment by ridiculousfish** [comment]
Ok this is fixed for real in d5d09c993e1c569ec4d5ba4cb24ff706798382eb. Thanks for staying with me as we debugged this!
